### PR TITLE
Codechange: make TimerGameCalendar Date and Year types strongly typed

### PR DESCRIPTION
**Pull request by TrueBrain**
## Motivation / Problem

With #10700 on the horizon, I am a bit afraid we will mix up Calendar with Economy, and get some weird buggy behaviour because of it. So I was wondering: can we make Date strongly typed, to avoid that issue.

This PR is a proof-of-concept for that. Curious what developers think.

And of course, by making it strongly typed, there is no actual functional change. It is just meant to assist the developer in not making silly mistakes. This was already fruitful, looking at all the PRs that came from this PR, with weird mistakes and boo-boos we made in our code-base.

## Description

I went for a slightly different approach than `TileIndex`, as that requires defining a struct. I was more interested in using `using` to create a new type. Got some inspiration from https://github.com/anthonywilliams/strong_typedef to make that happen.

Initially, I made the conversion to `int32` implicit (so it auto-cases to `int32` if it wants to). This had as nasty side-result that you could assign a `Year` to a `Date` and the other way around. As the compiler, as clever as it is, casted the one first to an int, to assign it to the other. This is kinda defeating the purpose of having strong-types to prevent people mixing up Calendar and Economy.

So, making it explicit is the way to go. Sadly, that showed tons of places in our code where we take some ... freedom in terms of using a date/year to use as integer. The weirdest things. A lot of small PRs came from this to address most of these.

## Limitations

Although performance shouldn't be impacted, as this is just compiler-hints etc, it is not a guarantee that there won't be any if used wrongly. Especially when disabling all optimization, some performance degradation is to be expected.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by TrueBrain** [comment]
I shifted gears, and went for a "I want to use `using` to define a strongly typed type`. I got some inspiration, and that is easily doable. Means we don't need to define a struct to create a new type, but we can just use `using`.

The trick here is that to avoid 2 `NewStrongTypedef` of the same base-type being equal, that we tag them with an unique tag. This works wonderful.

Sadly, I noticed that the compiler was sneaky converting everything back and forth between `int32` for us, so you could assign a year to a date. To solve this, I made the conversion to `int32` explicit, and boy, this showed a lot of places were we just depend on that. Not surprising, given we came from a world where you could. But plenty of places that are ... weird. Like `% 7` on the date.

Anyway, much more interested in the overall idea of this PR, if we like defining such type this strongly. Does it actually help?

**Comment by TrueBrain** [comment]
Another update. I decided to go for Mixins, after the idea of https://github.com/anthonywilliams/strong_typedef.

What is nice about this, is that it means `TileIndex` can be converted too, and we deduplicate defining a new type by a lot.
The downside is that the line to define a type becomes a bit lengthy :)

**Comment by 2TallTyler** [comment]
@TrueBrain A lot of this PR is way over my head with templating and casting, but you are absolutely right that we'll need this before #10700. Maybe some more experienced developers can weigh in on this? I would be eternally grateful. ❤️

**Comment by 2TallTyler** [comment]
The same way you make `SetDParam()` take strong-typed date arguments, it might be helpful (if that's possible) to do the same with settings. That could save a lot of casting, I think.

File: src/network/network_server.cpp [hunk 1]
```diff
@@ -1838,7 +1838,7 @@ static IntervalTimer<TimerGameCalendar> _network_daily({TimerGameCalendar::DAY,
 	if (!_network_server) return;
 
 	NetworkAdminUpdate(ADMIN_FREQUENCY_DAILY);
-	if ((TimerGameCalendar::date % 7) == 3) NetworkAdminUpdate(ADMIN_FREQUENCY_WEEKLY);
+	if (((int32_t)TimerGameCalendar::date % 7) == 3) NetworkAdminUpdate(ADMIN_FREQUENCY_WEEKLY);
```
Comment (TrueBrain) [inline src/network/network_server.cpp]: TODO: this should be fixed more properly, by adding `WEEK` and `QUARTER` similar as `DAY` in `TimerGameCalendar`, so a timer could be used here too.

File: src/saveload/oldloader_sl.cpp [hunk 2]
```diff
@@ -416,7 +416,7 @@ static bool FixTTOEngines()
 			if (TimerGameCalendar::date >= e->intro_date && HasBit(e->info.climates, 0)) {
 				e->flags |= ENGINE_AVAILABLE;
 				e->company_avail = MAX_UVALUE(CompanyMask);
-				e->age = TimerGameCalendar::date > e->intro_date ? (TimerGameCalendar::date - e->intro_date) / 30 : 0;
+				e->age = TimerGameCalendar::date > e->intro_date ? (int32_t)(TimerGameCalendar::date - e->intro_date) / 30 : 0;
```
Comment (TrueBrain) [inline src/saveload/oldloader_sl.cpp]: We kinda need DateToMonth or something here too. The `/ 30` is unclear as it is.

Edit: or just unmagicify the `30` with a constant. Might be sufficient too.

File: src/table/settings/world_settings.ini [hunk 3]
```diff
@@ -135,9 +135,9 @@ var      = game_creation.ending_year
 type     = SLE_INT32
 from     = SLV_ENDING_YEAR
 flags    = SF_GUI_0_IS_SPECIAL
-def      = DEF_END_YEAR
-min      = MIN_YEAR
-max      = MAX_YEAR - 1
+def      = (int32_t)DEF_END_YEAR
+min      = (int32_t)MIN_YEAR
+max      = (int32_t)MAX_YEAR - 1
```
Comment (TrueBrain) [inline src/table/settings/world_settings.ini]: See if we can render these casts automatically

File: src/saveload/oldloader_sl.cpp (hunk 2 above)
Comment (TrueBrain) [reply src/saveload/oldloader_sl.cpp]: Not for this PR. It needs a cast either way, so this is fine.

File: src/core/math_func.hpp [hunk 4]
```diff
@@ -254,10 +261,14 @@ static inline bool IsInsideBS(const T x, const size_t base, const size_t size)
  * @param max The maximum of the interval
  * @see IsInsideBS()
  */
-template <typename T>
+template<typename T, std::enable_if_t<std::disjunction_v<std::is_convertible<T, size_t>, std::is_base_of<StrongTypedefBase, T>>, int> = 0>
```
Comment (2TallTyler) [inline src/core/math_func.hpp]:
```suggestion
template <typename T, std::enable_if_t<std::disjunction_v<std::is_convertible<T, size_t>, std::is_base_of<StrongTypedefBase, T>>, int> = 0>
```
Other uses in this file have a space, while cppreference has no space. We could at least be consistent with ourselves. 😄

File: src/landscape.cpp [hunk 5]
```diff
@@ -787,7 +787,7 @@ void RunTileLoop()
 		_tile_type_procs[GetTileType(tile)]->tile_loop_proc(tile);
 
 		/* Get the next tile in sequence using a Galois LFSR. */
-		tile = (tile >> 1) ^ (-(int32_t)(tile & 1) & feedback);
+		tile = (static_cast<uint32_t>(tile) >> 1) ^ (-(int32_t)(static_cast<uint32_t>(tile) & 1) & feedback);
```
Comment (2TallTyler) [inline src/landscape.cpp:790]: The changes in this file deal with tiles, not dates. Should it be a separate commit or maybe a separate PR?

(Same goes for changes in `newgrf_house.cpp`, `newgrf_industries.cpp`, `newgrf_town.cpp`, `town_cmd.cpp`, `yapf_node.hpp`, and three files in `saveload/*`, just to make a nice list for you of what I found. 😉)

File: src/newgrf_engine.cpp [hunk 6]
```diff
@@ -597,7 +597,7 @@ static uint32_t VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *objec
 			}
 
 		case 0x4B: // Long date of last service
-			return v->date_of_last_service_newgrf;
+			return static_cast<int32_t>(v->date_of_last_service);
```
Comment (2TallTyler) [inline src/newgrf_engine.cpp]:
```suggestion
			return static_cast<int32_t>(v->date_of_last_service_newgrf);
```

File: src/script/api/script_basestation.cpp [hunk 7]
```diff
@@ -63,5 +63,5 @@
 {
 	if (!IsValidBaseStation(station_id)) return ScriptDate::DATE_INVALID;
 
-	return (ScriptDate::Date)::BaseStation::Get(station_id)->build_date;
+	return (ScriptDate::Date)(int32_t)::BaseStation::Get(station_id)->build_date;
```
Comment (2TallTyler) [inline src/script/api/script_basestation.cpp:66]: It feels weird to cast to an `int32_t` and then immediately to a `ScriptDate::Date`. Can we not do this in one step? (Honest question, I have no idea how ScriptDates work, nor casting in general. 😅)

File: src/timer/timer_game_calendar.h [hunk 8]
```diff
@@ -76,10 +77,12 @@ class TimerGameCalendar {
...
 	};
 
-	using Date = int32_t; ///< The type to store our dates in
+	/** The type to store our dates in. */
+	using Date = StrongType::Typedef<int32_t, struct DateTag, StrongType::Explicit, StrongType::Compare, StrongType::Integer>;
+
 	using DateFract = uint16_t; ///< The fraction of a date we're in, i.e. the number of ticks since the last date changeover
 
-	using Year = int32_t; ///< Type for the year, note: 0 based, i.e. starts at the year 0.
+	using Year = StrongType::Typedef<int32_t, struct YearTag, StrongType::Explicit, StrongType::Compare, StrongType::Integer>; ///< Type for the year, note: 0 based, i.e. starts at the year 0.
```
Comment (2TallTyler) [inline src/timer/timer_game_calendar.h]:
```suggestion
	/** Type for the year, note: 0 based, i.e. starts at the year 0. */
	using Year = StrongType::Typedef<int32_t, struct YearTag, StrongType::Explicit, StrongType::Compare, StrongType::Integer>;
```
If a block comment is good enough for Date, it's good enough for this. 😉 

...is your OCD contagious? 😛

File: src/script/api/script_basestation.cpp (hunk 7 above)
Comment (TrueBrain) [reply src/script/api/script_basestation.cpp:66]: Sadly, no. The first case changes the StrongTypeDef to an integer, the second to the Script-specific type. It is a bit annoying, but it is what it is. Endless casts in scripts are just .. what they are :P

File: src/landscape.cpp (hunk 5 above)
Comment (TrueBrain) [reply src/landscape.cpp:790]: No, sadly, these where it runs an operator on the StrongTypeDef, like `&`, `>>`, etc, has to be done here. Before this PR, that was a valid operator. But I think it has very little definition .. what type does `date >> 1` return .. so I revoked access to all those kind of operators.

As a downside, it means that the `TileIndex` also no longer can use them. And those you listed. So they have to be in this PR!

Comment (2TallTyler) [reply src/landscape.cpp:790]: Okay! Maybe a separate commit, then? ("No" is a valid answer. 😃)

Comment (TrueBrain) [reply src/landscape.cpp:790]: That would be really really difficult to make that a sane and compilable commit :) I rather keep it like this. It is only a few lines :)

**Review by 2TallTyler — DISMISSED** [review]
I don't understand template stuff well enough to review it, but the casting all looks correct to me. 

If someone else reviews the template stuff, I can continue working on #10700 tomorrow! Thanks again for taking this on! 😄

File: src/core/strong_typedef_type.hpp [hunk 9]
```diff
@@ -10,74 +10,199 @@
...
+			friend constexpr bool operator ==(const TType &lhs, TCompatibleType rhs) { return lhs.value == (TBaseType)rhs; }
+			friend constexpr bool operator !=(const TType &lhs, TCompatibleType rhs) { return lhs.value != (TBaseType)rhs; }
+
+			friend constexpr bool operator <=(const TType &lhs, TCompatibleType rhs) { return lhs.value <= (TBaseType)rhs; }
+			friend constexpr bool operator <(const TType &lhs, TCompatibleType rhs) { return lhs.value < (TBaseType)rhs; }
+			friend constexpr bool operator >=(const TType &lhs, TCompatibleType rhs) { return lhs.value >= (TBaseType)rhs; }
+			friend constexpr bool operator >(const TType &lhs, TCompatibleType rhs) { return lhs.value > (TBaseType)rhs; }
+
+			friend constexpr TType operator +(const TType &lhs, TCompatibleType rhs) { return { (TBaseType)(lhs.value + rhs) }; }
+			friend constexpr TType operator -(const TType &lhs, TCompatibleType rhs) { return { (TBaseType)(lhs.value - rhs) }; }
```
Comment (michicc) [inline src/core/strong_typedef_type.hpp]: The templates in general look as reasonable as they can I think.

For this specific area, was the C-style cast an explicit choice, i.e. do you want the `reinterpret_cast` case to included?

**Review by 2TallTyler — APPROVED** [review]

**Pull request merged**